scanners/seo.py

```diff
@@ -0,0 +1,146 @@
+import logging
+import requests
+
+from bs4 import BeautifulSoup
+
+"""
+Fairly simple scanner that makes a few checks for SEO and
+search indexing readiness. It will attempt to determine:
+* Does a robots.txt file exist at the expected location (/robots.txt)
+* If a sitemap.xml file exists
+  * If so, how many URLs are in it?
+  * How many of those URLs are PDF files?
+* Does a `main` element exist on pages
+* Does OG data exist (particularly date)
+
+Further, by comparing the site root with a second page (/privacy) it
+will compare the two pages to determine if the title and descriptions
+appear to be unique.
+
+Possible future scope:
+* crawl delay (how do we determine?)
+* total urls (set of nav + sitemap maybe?)
+* builtwith (requires paid API access)
+* expanded checking (grab some set of URLs from Nav and look at them, too)
+
+"""
+
+# Set a default number of workers for a particular scan type.
+# Overridden by a --workers flag. XXX not actually overridden?
+workers = 50
+
+# This is the initial list of pages that we will be checking.
+pages = [
+    "/",
+    "/privacy",
+]
+
+
+# Optional one-time initialization for all scans.
+# If defined, any data returned will be passed to every scan instance and used
+# to update the environment dict for that instance
+# Will halt scan execution if it returns False or raises an exception.
+#
+# Run locally.
+def init(environment: dict, options: dict) -> dict:
+    logging.debug("Init function.")
+    return {'pages': pages}
+
+
+# Required scan function. This is the meat of the scanner, where things
+# that use the network or are otherwise expensive would go.
+# Runs locally or in the cloud (Lambda).
+def scan(domain: str, environment: dict, options: dict) -> dict:
+    logging.debug("Scan function called with options: %s" % options)
+
+    # Initialize results with a quick check for the presence of robots.txt.
+    # Because we're not reading robots.txt, we only need the status code.
+    # We could put these in `pages` but since we have specialized needs for each
+    # and no need to iterate through them, I don't want to.
+    results = {
+        'robots': requests.get("https://" + domain + '/robots.txt').status_code,
```

Comment:
Thoughts on using `urllib.parse.urljoin` or `path.join` to build the urls? That way we can avoid having to keep track of `/`s.

Reply:
🤷 No real opinion either way. I was just trying to be consistent with how they're constructed elsewhere (at least in the example I started from). If you feel strongly about it, I'm happy to change it. I do agree, the presence or absence of `/` is pretty easy to get wrong.

Reply:
Sounds good. I think in the spirit of consistency we keep it as is.